### ISSUE TITLE
rect_size of Control node with tool script is (0, 0) in _ready

### ISSUE THREAD

**Issue by DualMatrix** [godotengine/godot#20896]
**Godot version:**
3.0.6

**Issue description:**

The value of rect_size of a Control node is (0, 0) in _ready when it is called in a tool script.

**Comment by isaacremnant** [comment]
I'm getting this issues with all controls. It seems that when the anchors are not equal, rect_size is improperly computed in _ready (but not afterwards in process!)...

In ready, rect_size is (0, 0) on the left and (169, 220) on the right:
![untitled](https://user-images.githubusercontent.com/35355161/44887229-165d7f00-ac99-11e8-9460-6f58f260f8d3.png)

**Comment by isaacremnant** [comment]
Just an update because I think I understand where this is coming from. Looking at `control.cpp`, `rect_size` needs to be computed based on the parent's `rect_size`. Since children get initialized before their parents, calling `get_parent().rect_size` will always return (0,0) in `_ready()`.

This means that in `_ready()`, `rect_size` can only be reported as `Vector2(margin_right-margin_left, margin_bottom-margin_top)`. This is not actually a problem when all four anchors are 0.

I don't think this is actually a bug, but a limitation of the initialization order. It's still pretty annoying though, so I'd at least like an actual error when accessing `rect_size` in `_ready`. Maybe `rect_size` could be NULL until it's properly initialized? Workaround is something like
```
func _ready():
    call_deferred("READY")
func READY():
    do something with now properly initialize rect_size
```

**Comment by vnen** [comment]
You don't need to create another function, you can use `yield(get_tree(), "idle_frame")` too.

I don't see this is a bug, it's just how controls work. It might need to be better documented.

**Comment by avencherus** [comment]
I would agree it probably needs to be documented that these aren't updated immediately.  Also worth noting that they don't update when the control is not visible, and only after the frame when they're first made visible again.

**Comment by girng** [comment]
when using the `idle_frame` trick to re-adjust the label / control, there is a slight artifact/flicker. which is infinitesimally small, but noticeable. especially if you have a style attached to the label. it should be updated internally in the renderer

**Comment by isaacremnant** [comment]
> artifact/flicker
It's exactly one frame because you're drawing as soon as the control is ready. Just hide() it in _ready then show() it after the update.

**Comment by girng** [comment]
> It's exactly one frame because you're drawing as soon as the control is ready. Just hide() it in _ready then show() it after the update.

oh, i mean after _ready, when adjusting a label's size. sorry got this issue confused with my issue

**Comment by aaronfranke** [comment]
Would it be possible to somehow block access to `rect_size` only in `_ready()`?

**Comment by Shadowblitz16** [comment]
This is a bug rect_size is clearly 126x126 on my end but when I use rect_size it returns the same thing as min_rect_size.
Its also not in any containers.
This is in ready where the node should be already initialized 
![image](https://user-images.githubusercontent.com/2453707/146657887-f9880312-7006-45b8-8a44-39ad022cba8d.png)
```
extends ColorRect

export var value : float
export var speed : float

var maxValue    : float
var maxValueBar : float

func _ready():
	maxValue    = value
	maxValueBar = $ColorRectFG.rect_min.x
	yield(get_tree().create_timer(4), "timeout")
	damage(maxValue)
	
func damage(amount:float)->void:
	var newSizeX = range_lerp(maxValue, 0, maxValue, 0, maxValueBar)
	print(maxValueBar)
	print(newSizeX)
	
	$Tween.interpolate_property($ColorRectFG, "rect_size", rect_size, Vector2(newSizeX, $ColorRectFG.rect_size.y), 0.0  , Tween.TRANS_LINEAR, Tween.EASE_IN)
	$Tween.start()
	
	$Tween.interpolate_property($ColorRectMG, "rect_size", rect_size, Vector2(newSizeX, $ColorRectFG.rect_size.y), speed, Tween.TRANS_LINEAR, Tween.EASE_IN)
	$Tween.start()
	
```

Please stop always assuming things are not bugs. Godot is filled with bugs that the Dev's ignore and then call it a documentation error because they don't want to fix them or make Godot better

**Comment by Shadowblitz16** [comment]
I can confirm this is a bug...
```
extends ColorRect

export var value : float
export var speed : float

var maxValue    : float
var maxValueBar : float

func _ready():
	maxValue    = value
	maxValueBar = $ColorRectFG.rect_size.x
	damage(maxValue)
	
func damage(amount:float)->void:
	var newSizeX = range_lerp(maxValue, 0, maxValue, 0, maxValueBar)
	print(maxValueBar)
	print(newSizeX)
	
	$Tween.interpolate_property($ColorRectFG, "rect_size", rect_size, Vector2(newSizeX, $ColorRectFG.rect_size.y), 0.0  , Tween.TRANS_LINEAR, Tween.EASE_IN)
	$Tween.start()
	
	$Tween.interpolate_property($ColorRectMG, "rect_size", rect_size, Vector2(newSizeX, $ColorRectFG.rect_size.y), speed, Tween.TRANS_LINEAR, Tween.EASE_IN)
	$Tween.start()
```
```
--- Debugging process started ---
Godot Engine v3.4.1.stable.official.aa1b95889 - https://godotengine.org
OpenGL ES 3.0 Renderer: AMD RENOIR (DRM 3.40.0, 5.11.0-43-generic, LLVM 12.0.0)
OpenGL ES Batching: ON
 
0
0
```
![image](https://user-images.githubusercontent.com/2453707/146657958-b66e90de-198b-4351-b8c3-f702cd290ca4.png)

The min_rect_size is 0 which is what its pulling from.

**Comment by wallyfoo** [comment]
I ran into this today as well. 

> yield(get_tree(), "idle_frame")

This feels hacky, though this is what I'm going with at the moment to get the rect_size of a Label node after I assign its text in _ready(). I can understand that the ready pass happens on the same tick, though wouldn't necessarily update, but if I were to offer a solution, perhaps explicitly allowing an update() to the control node would let us return the size values?

What I ended up landing on is a utility function I can call at will that helps with a bit of a flicker artifact:

```
func size_update_hack() -> void:
	hide()
	yield(get_tree(), "idle_frame") 
	show()
```
